### PR TITLE
Add tool for preparing local GeoIp database service

### PR DESCRIPTION
**Pull request by probakowski**
Air-gapped environments can't simply use GeoIp database service provided by Infra, so they have to either use proxy or recreate similar service themselves.
This PR adds tool to make this process easier. Basic workflow is:

- download databases from MaxMind site to single directory (either `.mmdb` files or gzipped tarballs with `.tgz` suffix)
- run the tool with `$ES_PATH/bin/elasticsearch-geoip -s directory/to/use [-t target/directory]` 
- serve static files from that directory (for example with `docker run -v directory/to/use:/usr/share/nginx/html:ro nginx`
- use server above as endpoint for `GeoIpDownloader` (`geoip.downloader.endpoint` setting)
- to update new databases simply put new files in directory and run the tool again

This change also adds support for relative paths in overview json because the cli tool doesn't know about the address it would be served under.

Relates to https://github.com/elastic/elasticsearch/issues/68920

### CODE REVIEW

File: distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import static java.nio.file.StandardOpenOption.TRUNCATE_EXISTING;
+
+public class GeoIpCli extends Command {
+
+    private static final byte[] EMPTY_BUF = new byte[512];
+    private final OptionSpec<String> directoryArg;
+
+    public GeoIpCli() {
+        super("A CLI tool to prepare local GeoIp database service", () -> {});
+        directoryArg = parser.acceptsAll(Arrays.asList("d", "directory"), "Directory to process").withRequiredArg().defaultsTo(".");
```
Comment (martijnvg) [inline distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java:47]: I don't think we should default to `.` here. I would prefer that the help page gets printed if no arguments are provided to this command.

File: distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    @SuppressForbidden(reason = "file arg for cli")
+    private Path getPath(String file) {
+        return PathUtils.get(file);
+    }
+
+    private void packDatabasesToTgz(Terminal terminal, Path directory) throws IOException {
+        List<Path> toPack = Files.list(directory).filter(p -> p.getFileName().toString().endsWith(".mmdb")).collect(Collectors.toList());
+        for (Path path : toPack) {
+            String fileName = path.getFileName().toString();
+            Path compressedPath = path.resolveSibling(fileName.replaceAll("mmdb$", "") + "tgz");
```
Comment (martijnvg) [inline distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java:66]: Should we allow for an additional argument that allows specify the target directory (which then defaults to the source directory)? The `d` argument is now both the source and target directory.

**Review by martijnvg — COMMENTED** [review]
Thanks @probakowski, I left a few questions.

File: distribution/tools/geoip-cli/src/test/java/org/elasticsearch/geoip/GeoIpCliTests.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+ */
+
+package org.elasticsearch.geoip;
+
+import joptsimple.OptionSet;
+import joptsimple.OptionSpec;
+import org.apache.lucene.util.LuceneTestCase;
+import org.elasticsearch.cli.MockTerminal;
+import org.elasticsearch.common.SuppressForbidden;
+import org.junit.Before;
```
[the remark this answers is not in the document]
Comment (probakowski) [reply distribution/tools/geoip-cli/src/test/java/org/elasticsearch/geoip/GeoIpCliTests.java]: removed

File: distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java (hunk 1 above)
Comment (probakowski) [reply distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java]: That makes sense, I've changed it

File: distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java (hunk 2 above)
Comment (probakowski) [reply distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java]: Yes, this is good idea, I've changed `d[ir]` to `s[ource]` and `t[arget]` pair and added tests for using 2 different dirs and defaulting to source

File: distribution/tools/geoip-cli/build.gradle [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  compileOnly project(":server")
+  compileOnly project(":libs:elasticsearch-cli")
+  compileOnly project(":libs:elasticsearch-x-content")
+  testImplementation project(":test:framework")
+}
+
+tasks.named("test").configure {
+  // TODO: find a way to add permissions for the tests in this module
+  systemProperty 'tests.security.manager', 'false'
+}
```
[the remark this answers is not in the document]
Comment (probakowski) [reply distribution/tools/geoip-cli/build.gradle]: You're right, I thought it was necessary but it turns out it's not needed. I've removed it

**Comment by probakowski** [comment]
@elasticmachine update branch

File: distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            return;
+        }
+        List<Path> toCopy = Files.list(source).filter(p -> p.getFileName().toString().endsWith(".tgz")).collect(Collectors.toList());
+        for (Path path : toCopy) {
+            Files.copy(path, target.resolve(path.getFileName()), StandardCopyOption.REPLACE_EXISTING);
+        }
+    }
+
+    private void packDatabasesToTgz(Terminal terminal, Path source, Path target) throws IOException {
+        List<Path> toPack = Files.list(source).filter(p -> p.getFileName().toString().endsWith(".mmdb")).collect(Collectors.toList());
```
Comment (martijnvg) [inline distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java:82]: Should we also pack txt files? There could be license and notice files?

File: distribution/tools/geoip-cli/src/test/java/org/elasticsearch/geoip/GeoIpCliTests.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            for (Map<String, String> map : list) {
+                assertThat(map, hasKey("md5_hash"));
+                assertThat(map, hasKey("updated"));
+            }
+        }
+    }
+
+    private void verifyTarball(Map<String, byte[]> data) throws Exception {
+        // skip tarball verifications on Windows, no tar utility there
+        if (WINDOWS) {
```
Comment (martijnvg) [inline distribution/tools/geoip-cli/src/test/java/org/elasticsearch/geoip/GeoIpCliTests.java:127]: We could add a tar library as a test dependency and then use that to verify the tarball. That way we don't have to rely on the `tar` command line utility to be available on the machine that runs the tests.

**Review by martijnvg — COMMENTED** [review]
👍 I left two more questions.

Ideally we also would want to have a qa module that tests this entire workflow.
This qa module would prior to starting the test cluster, execute the `elasticsearch-geoip` and then start a webserver (built-in jdk webserver?). I don't think this is required
for getting this pr merged, but maybe after we get the docs in, we can investigate this?

File: distribution/tools/geoip-cli/src/test/java/org/elasticsearch/geoip/GeoIpCliTests.java (hunk 6 above)
Comment (probakowski) [reply distribution/tools/geoip-cli/src/test/java/org/elasticsearch/geoip/GeoIpCliTests.java]: Turns out we already have apache common-compress in test framework, I've used it and it makes code much cleaner. Thanks for suggestion!

File: distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java (hunk 5 above)
Comment (probakowski) [reply distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java:82]: I don't think we have to, if users downloaded database from MaxMind they already had a chance to check LICENSE.txt, COPYRIGHTS.txt and so on, no need to copy it over to the tgz and ingest nodes. WDYT?

**Comment by probakowski** [comment]
Thanks for the review @martijnvg, I think qa module you mentioned is great idea, I'd rather add it in separate PR to keep this one smaller.

File: distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java (hunk 5 above)
Comment (martijnvg) [reply distribution/tools/geoip-cli/src/main/java/org/elasticsearch/geoip/GeoIpCli.java:82]: That is true.

**Comment by martijnvg** [comment]
> I think qa module you mentioned is great idea, I'd rather add it in separate PR to keep this one smaller.

Agreed. Let's first focus on getting the docs changes for geoipv2.

**Review by martijnvg — APPROVED** [review]
LGTM

**Pull request merged**